vltt/assignment4/Solution_4.cpp

```diff
@@ -0,0 +1,138 @@
+#include <iostream>
+#include <algorithm>
+#include <iomanip>
+#include <vector>
+#include <map>
+#include <unordered_set>
+#include <set>
+#include <string>
+#include <memory>
+
+#define EXPECT_EQ(expected, actual)                                       \
+  if ((expected) != (actual)) {                                           \
+    std::cout << __FILE__ << ":" << __LINE__ << " Failed: "               \
+              << "Statement (" << #actual << ") is " << actual            \
+              << ", expected " << (expected) << " (" << #expected << ")." \
+              << std::endl;                                               \
+  } else {                                                                \
+    std::cout << __FILE__ << ":" << __LINE__ << " True " << std::endl;  \
+  }
+
+template <class T>
+inline std::ostream& operator<<(std::ostream& os, const std::set<T>& s) {
+	os << "{";
+	for (auto &it : s) {
+		os << it;
+		std::cout << ", ";
+	}
+	os << "}";
+	return os;
+}
+
+
+void make_set(int num_island, std::vector<int> &parent, std::vector<int> &rank) {
+	parent[num_island] = num_island;
+	rank[num_island] = 0;
+}
+
+int find_set(int num_island, std::vector<int> &parent) {
+	if (num_island == parent[num_island])
+		return num_island;
+	return parent[num_island] = find_set(parent[num_island], parent);
+}
+
+void union_sets(int num_island_a, int num_island_b, std::vector<int> &parent, std::vector<int> &rank) {
+	num_island_a = find_set(num_island_a, parent);
+	num_island_b = find_set(num_island_b, parent);
+	if (num_island_a != num_island_b) {
+		if (rank[num_island_a] < rank[num_island_b])
+			std::swap(num_island_a, num_island_b);
+		parent[num_island_b] = num_island_a;
+		if (rank[num_island_a] == rank[num_island_b])
+			++rank[num_island_a];
+	}
+}
+
+//disjoint-set-union (with path compression and union by rank)
+int number_of_islands(int n, int m, std::vector<std::vector<bool> > &location_of_isl){
```

Comment:
It's not a bad idea to be more 'defensive' when writing code.

This is especially useful when you're preparing for interviews. Your code should be bullet-proof :) 
In this case, you're just accepting that the given n and m are the actual dimensions of your map. What if they're not? What if I pass -20 for n? What if I set n and m to 5, but the map is empty? 

A better way to deal with this, would be to just calculate n and m within this function. Sure, this would mean that you have to do another O(N*M) to calculate the sizes, but often in the real world you need sacrifice speed for correctness and robustness. This is especially true when you're dealing with user input, or allow someone who's not you to call your function. Others might be easily confused by what you're asking from them, especially when you don't have comments describing what everything is or should be ;)